### ISSUE TITLE
./update fails due to virtualenv

### ISSUE THREAD

**Issue by malcolmwhat** [mcgill-robotics/compsys#42]
As per @anassinator s suggestion on #41 I updated my local compsys repo and ran `./update`.

In the pip2 update step, I get the following:
```
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-ab3o7D/ipython/
```

**Comment by anassinator** [comment]
Looks like you have some out of date `pip` or `setuptools`. Try upgrading them manually for both `python2` and `python3`.

**Comment by foochuanwei** [comment]
@malcolmwhat Did you get the line `Found mismatch in default pip. Correcting...` after the error? If so, the cause of the failure may be that `python setup.py egg_info` is using the wrong version of python.

**Comment by malcolmwhat** [comment]
Ok so I fixed it, I did not get that message @foochuanwei, however it looks like that was the error.
Here's the [full log](https://github.com/mcgill-robotics/compsys/files/1011782/log.txt) for the steps it took to fix it, but I'll highlight the important stuff:

Pip versions:

| Command Issued | Result | Comment |
| ----------------------- | --------- | ------------- |
|   `pip --version`  |   `pip 9.0.1 from /home/mw/.local/lib/python2.7/site-packages (python 2.7)`  |   Fine |
| `pip3 --version` | `pip 9.0.1 from /home/mw/.local/lib/python3.5/site-packages (python 3.5)` | Fine |
| `sudo pip --version` | `pip 9.0.1 from /home/mw/.local/lib/python3.5/site-packages (python 3.5)` | Default version of `pip` is `pip3` |
| `sudo pip2 --version` | `pip 8.1.1 from /usr/lib/python2.7/dist-packages (python 2.7)` | Outdated |

Anyway once I saw that root's `pip2` was outdated I upgraded it (which also set my `pip` symlink to point to pip2) and re-ran the update script. Then I got 
```bash
Updating pip2 dependencies...
    DEPRECATION: Uninstalling a distutils installed project (enum34) has been deprecated and will be removed in a future version. This is due to the fact that uninstalling a distutils project will only partially uninstall the project.
    DEPRECATION: Uninstalling a distutils installed project (ptyprocess) has been deprecated and will be removed in a future version. This is due to the fact that uninstalling a distutils project will only partially uninstall the project.
    DEPRECATION: Uninstalling a distutils installed project (pexpect) has been deprecated and will be removed in a future version. This is due to the fact that uninstalling a distutils project will only partially uninstall the project.
    DEPRECATION: Uninstalling a distutils installed project (PyYAML) has been deprecated and will be removed in a future version. This is due to the fact that uninstalling a distutils project will only partially uninstall the project.
    DEPRECATION: Uninstalling a distutils installed project (docutils) has been deprecated and will be removed in a future version. This is due to the fact that uninstalling a distutils project will only partially uninstall the project.
Updating vim plugins...
Done.
```
So it succeeded. You can check the logs for more detail.

It looks like the issue is that on [this line](https://github.com/mcgill-robotics/compsys/blob/master/update#L33) you check the user's pip version instead of root's pip version, but the update step 2 lines above is done with super user privileges.

**Comment by malcolmwhat** [comment]
We can easily change that line to 
```bash
if [[ $(sudo -H pip --version) != $(sudo -H pip2 --version) ]]; then
```
but then the question is how do we update the symlinks. Not sure if this is odd or not but I have two distinct pip symlinks: one at `/usr/bin/pip` and one at `/usr/local/bin/pip`, so we would have to delete both of those, and redo the symlinks there.

**Comment by foochuanwei** [comment]
@malcolmwhat Having both is normal. `/usr/bin/pip` is created when installing `python-pip` using `apt-get`. `/usr/local/bin/pip` is created when installing `pip` using the `pip` that was installed using `apt-get`.

**Comment by foochuanwei** [comment]
I'm not very sure, but do you think that there should be a hash -r before this line? Could be something we overlooked. The path to the pip executable may have changed and bash may not know about it.

**Comment by anassinator** [comment]
This case is very specific, so I'm not sure I would even bother fixing this. This issue is simply due to the fact that the user's `pip` is different from root's `pip`. I'm guessing you have a virtualenv or something.

I'm also pretty sure a `sudo ./update` would fix this.

**Comment by malcolmwhat** [comment]
I do have a virtualenv, but the issue is most likely the result of an incorrect initial setup of said environment. 

I agree, it is pretty specific. However we should address the fact that the [pip version fix in the update script](https://github.com/mcgill-robotics/compsys/blob/master/update#L33-L40) does not work as expected. The adjustment of the symlinks will adjust the symlinks in the user's space (assuming this is run without root privileges):

```bash
PIP_PATH=$(which pip)
```
will set the `PIP_PATH` variable to something like `/home/user/.local/bin/pip`, and `/home/user/.local/bin/pip2` for the `PIP2_PATH`. Then the adjustment of symlinks will only impact the user's version of `pip` and not root's (even if we didn't have a mismatch in the system before, this would create one, as the user's version of pip would become `pip2` but root's would remain `pip3`).

Running with superuser privileges would update the symlinks in `/usr/local/bin` but not in `/usr/bin`. Although it's possible that this would not have a negative impact, we should still be sure.

**Comment by anassinator** [comment]
I guess we actually only care about `sudo pip` since that's what `rosdep` uses, so I'll grant you that. Fixed as per suggestion in #43
